system/jlib/jfile.cpp

```diff
@@ -7106,3 +7109,230 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
     else
         return linuxFileSystemProperties;
 }
+
+
+#ifndef _WIN32
+class CFileEventWatcher : public CInterfaceOf<IFileEventWatcher>, implements IThreaded
+{
+    class CMonitorItem : public CInterface
+    {
+    public:
+        CMonitorItem(const char *_filename, int _watchFd, FileWatchEvents _events) : filename(_filename), watchFd(_watchFd), events(_events)
+        {
+        }
+        std::string filename;
+        int watchFd = -1;
+        FileWatchEvents events = FileWatchEvents::none;
+        bool pendingRemoval = false;
+    };
+    int inotifyFd = -1;
+    int pipefd[2] = {-1, -1};
+    pollfd toPoll[2];
+    std::vector<Owned<CMonitorItem>> monitoredFiles;
+    CriticalSection crit;
+    FileWatchFunc callback;
+    CThreaded threaded;
+    bool stopped = true;
+
+    void handle_events(int fd)
+    {
+        byte buf[4096] __attribute__ ((aligned(__alignof__(struct inotify_event))));
+
+        while (true)
+        {
+            ssize_t len = read(fd, buf, sizeof(buf));
+            if (len == -1 && errno != EAGAIN)
+                throw makeErrnoException("CFileEventWatcher::handle_events - read");
+
+            if (len <= 0)
+                break;
+
+            const byte *ptr = buf;
+            const struct inotify_event *event = (const struct inotify_event *) ptr;
+
+            do
+            {
+                std::string filename;
+                if (event->len) // if watching directory event->name is populated with file within that dir.
+                    filename = event->name;
+                else
+                {
+                    /* if watching a file, lookup filename from event handle
+                     * linear search, not worth hash table, not expected to be that many.
```

Comment:
Famous last words... is a hash table really that hard?

Reply:
have converted to use ht(s)